### ISSUE TITLE
Best Practices WebAPI Roles and Client Credentials Flow

### ISSUE THREAD

**Issue by S3KCentrifugal** [IdentityServer/IdentityServer3#2916]
This is basically a reposting of what is already in Github and in Stackoverflow. I'm not getting any responses, so it must be the way I am phrasing the question.

**Here's what is happening now...**
1. I have a WebAPI with methods secured by roles/claims. (They all come from Idsv3)
2. Clients using implicit flow make use of the API, with roles/claims in the access token of a user

**The goal...**
I now also need to make machine to machine from a background process to this same WebAPI. Different client, but same resource. This means client credentials flow, no user, no claims, no roles....

**Roadblock**
There is not a technical roadblock here, I could give a client using client credentials flow the roles/claims required for the WebAPI. However my OCD has gotten the better of me here. Based on my research it seems like the idea of associating roles/claims with a client might be a bad practice. 

I realize that this may not pertain directly to Idsv3, but as far as I am concerned the contributors to this repository are the most qualified subject matter experts. 

In my Stack Overflow question I also link to the Github discussions that triggered my confused. I feel like I am probably just over thinking the problem.
[https://stackoverflow.com/questions/37280538/webapi-mixing-implicit-flow-and-client-credentials-flow](https://stackoverflow.com/questions/37280538/webapi-mixing-implicit-flow-and-client-credentials-flow)

**Comment by chazt3n** [comment]
I'm digging on some startup slowness at the moment, I wish I had more time to discuss this. I may be missing the point here, however couldn't you have the Client run under a client/user/service account that will always be responsible for that Client, and give that guy the roles you need?

**Comment by S3KCentrifugal** [comment]
I could simply just use a long lived access token. There are several ways I can get this working against a WebAPI using role/claims authorization. Getting it to work isn't the problem though, I am strictly looking for best practices.

Based on all my readings (mostly from Dominick Baier and Scott Brady) the best approach for machine to machine communication is via Client Credentials Flow. Which of course only involves a client/secret, and with no subject/user. Also, that authorization should be done via scopes instead of claims for this flow. It seems like a great inconsistency to me since WebAPI methods may already exist and secured using roles/claims for existing implicit clients.

To my knowledge giving claims to client access tokens is not strictly forbidden in the protocol, so if @leastprivilege is advising against it then he must be working towards some kind of best practice. Or it is also possible I completely misunderstood his comments in other posts.

My solutions as I see them are...
1. Use long live access token, that will have a user/subject/claims
2. Assign claims/roles to clients using Client Credentials Flow
3. Create new APIs specifically for clients using different flows

I am not a huge fan of #3 since it means maintaining multiple sets of authorization logic for each API method.

If anyone wants to discuss further just let me know and I can pm my direct line. I've also got Skype/Skype for business.

**Comment by brockallen** [comment]
If there's no `sub` and there's a valid `scope` then the client has been authorized.

**Comment by S3KCentrifugal** [comment]
So here is my mistake (I think)...

I was building my WebAPI using the traditional claim/role authorization filters, my required scopes defined in startup for entire service and not per method. When creating a client that uses client credentials flow I expected it should work so long as the correct claims/roles are provided in token. After reading that clients should not be given claims/roles this effectively meant my API would not be available using Client Credentials flow.

Correct me if I am wrong... There is basically no best practice, no ability to place scope restrictions on controller methods that is native to .Net 4.5, and so a custom authorization filter is going to be required.

That said I do see a project Thinktecture.IdentityModel.45 that appears to already support this. However the project is flagged as deprecated, and I ran into an issue late last year where there were actually two version of Identity Server 3 namespace, "Thinktecture.IdentityServer" and "IdentityServer3". The former having been deprecated but documentation being hard to follow due to the ambiguous names. So I've almost intentionally avoided referencing anything in a Thinktecture namespace. So this was likely another mistake of mine. In the case of this specific project I can see now it is only deprecated due to the newer WebAPI V2 implementation.

**Comment by brockallen** [comment]
> When creating a client that uses client credentials flow I expected it should work so long as the correct claims/roles are provided in token.

And if you're too far down this road, you can make it work -- the `Client` configuration in IdSvr has the ability to assign claims.

> There is basically no best practice

There's rarely ever such a thing as a "beat practice" -- it depends on context.
